### PR TITLE
fix tb hparams logging

### LINKED ISSUE #2971: TensorBoardLogger not saving hparams without metrics
## 🐛 Bug

`log_hyperparams` for `TensorBoardLogger` saves no data with default `metrics=None`, only hparam entries/names show up in sidebar

### To Reproduce

```
import pytorch_lightning as pl
logger = pl.loggers.TensorBoardLogger("./test_logs")
test_dict = {"test":0}
logger.log_hyperparams(test_dict)                      ## no data saved
logger.log_hyperparams(test_dict, test_dict)           ## works
logger.log_metrics(test_dict)                          ## works
logger.experiment.add_hparams(test_dict, test_dict)    ## works but saves in a different events file
```

### Expected behavior
hparams data is saved and viewable via tensorboard with default args

### Proposed solution
For default hparams logging without metrics, add a placeholder metric? I can do a PR if this is appropriate.

### PR DESCRIPTION
**Pull request by s-rog**
## What does this PR do?

allows the default call to `log_hyperparams` to actually log hparams since it's called without a metric
also changes the method of logging to the official method in pytorch docs (will revert if problematic)

Fixes #2971 #2406

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
@s-rog  make sure to ping me here, once this is ready to review :)

**Comment by s-rog** [comment]
@justusschock should the log_hparams metrics just be locked to scalar? since fit by default calls log_hparams without specifying metrics the user cannot change the key of the metric logged anyways. and the key cannot be updated by a subsequent call, only the value.

**Comment by justusschock** [comment]
I don't understand, why you cannot update the key?

**Comment by s-rog** [comment]
I see, for my own workflow I would want to see hparams from the get go, hence this implementation. Unless this is not the norm?

**Comment by justusschock** [comment]
yeah that's the tradeoff, why we had to make a decision. I'm not sure, what would be the best way here. Maybe log once, as soon as there is a metric dict passed and then just update these metrics?

**Comment by s-rog** [comment]
I have not tried separating logging the hparams themselves and the hparams metric, I'll test it out come monday. For default use cases that would still make hparams only be viewable at the end of training (if no metrics dict was passed during training and we add a default placeholder at the end).

This compromise works for me since I can still manually force hparams to show in the beginning, but it still may not be the best default mode. Any thoughts on this? @williamFalcon 

Also, regarding the tests I'm not sure why they are failing if the master passed. Currently they are failing because `self.log_metrics(metrics)` is decorated in tests forcing the optional step arg to be mandatory, but I did not change the call to log_metrics from master which is also missing step.

**Comment by s-rog** [comment]
> Maybe log once, as soon as there is a metric dict passed and then just update these metrics?

@justusschock logging the hparams without metrics just breaks logging (same as the current implementation)

**Comment by justusschock** [comment]
Mhm, this could be a way to go, although I don't like that we do different things for different loggers here

**Comment by s-rog** [comment]
yeah that's what I thought as well, but tb hparams logging is fundamentally different from other loggers

**Comment by justusschock** [comment]
Yeah, I see. Thoughts @Borda @williamFalcon @awaelchli @nateraw ?

**Comment by s-rog** [comment]
@justusschock since no one seems to have any input, I prefer the logger init arg approach. If you approve, I'll go ahead and work on that (possibly in a different PR).

**Comment by justusschock** [comment]
@s-rog yeah that'd be great!

**Comment by s-rog** [comment]
@justusschock should be mostly done, except updating docs

**Comment by justusschock** [comment]
So you habe a default metric now, that always shows up but the other ones are displayed as well?

**Review by Borda — APPROVED** [review]
LGTM

File: tests/loggers/test_all.py [hunk 1]
```diff
@@ -79,9 +79,16 @@ def log_metrics(self, metrics, step):
...
 
     log_metric_names = [(s, sorted(m.keys())) for s, m in logger.history]
-    assert log_metric_names == [(0, ['epoch', 'val_acc', 'val_loss']),
-                                (0, ['epoch', 'train_some_val']),
-                                (1, ['epoch', 'test_acc', 'test_loss'])]
+    if logger_class == TensorBoardLogger:
+        assert log_metric_names == [(0, ['hp_metric']),
+                                    (0, ['epoch', 'val_acc', 'val_loss']),
+                                    (0, ['epoch', 'train_some_val']),
+                                    (0, ['hp_metric']),
```
Comment (awaelchli) [inline tests/loggers/test_all.py:86]: If hp_metric is here twice on purpose, can you write a comment why?

Comment (s-rog) [reply tests/loggers/test_all.py:86]: because test calls log_hyperparams as well so it is logged twice (fit + test)

Comment (awaelchli) [reply tests/loggers/test_all.py:86]: ok

**Comment by pnmartinez** [comment]
> the value can yes, and it can be updated just by logging values to the hp_metric key, without calling log_hyperparams
> 
> if you want to log more than one metric for hparams then you can just disable the flag and call log_hyperparams manually

Hi! This indeed seems like a much-needed feature, but I can't find a good tutorial on the current state of its implementation.

**Can you elaborate** on how to overwrite the `hp_metric` placeholder? Can it be done like:

```
logger = TensorBoardLogger( 
                        tensor_log_dir, 
                        **default_hp_metric = 'val_loss'  # replaces "hp_metric" value on "metrics" dict?**
)  
```

I've tried that, but to no avail: **I keep getting the `hp_metric = -1`** all the time (having started logging on a fresh new folder and averything).

**Comment by jbohnslav** [comment]
I'm having an issue with this PR. I want to log multiple metrics, along with hyperparameters. However, calling [log_hyperparams](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.loggers.tensorboard.html#pytorch_lightning.loggers.tensorboard.TensorBoardLogger.log_hyperparams) with my metrics dictionary in addition to `lightning_module.log` duplicates all the values in TensorBoard. However, it's not possible to stop the calls to `lightning_module.log`, as that's how e.g. learning rate schedulers are written.

**Comment by s-rog** [comment]
@jbohnslav Do you have reproducible code that I can look at? Multiple hparams metrics work fine last I checked.

**Comment by jbohnslav** [comment]
Edit, I found a workaround just by using a separate key for my hyperparameter metrics. (e.g. calling `val_accuracy` `val_accuracy_hp` and passing that as a dictionary to `log_hyperparams`). If I have time in the future, I'll try to reproduce the double-logging and re-open.

**Comment by s-rog** [comment]
@jbohnslav 
I see what you mean, `log_hyperparams` does write its own metrics value so selecting the same key as other log calls would result in duplicate metrics. This is why the default metric key in this PR is not something like `val_loss`.

A little tip: if you have multiple hp metrics try `hp/{metric_name}` instead of `{metric_name}_hp` to group them in tb scalar view or collapse them altogether.

**Comment by FredrikM97** [comment]
I dont see the purpose of _default_hp_metric _. No documentation on how it should be used and when it is True the only value on tensorboard is -1. I did the same thing as @s-rog and created redundant plots. However I noted if I can log_hyperparams multiple times it will still stay on step 0 without moving. 

Adding a feature to change the step could be an idea, I am calling log_hyperparams on epoch_end since I want to log the validation and training loss and compare it to other models? Or link to existing plots if they exist instead of forcing to create new ones.

**Comment by s-rog** [comment]
log_hyperparams should only be called once. If you want to update the metric, only update the metric. This is a TB design choice and not a lightning one.

**Comment by FredrikM97** [comment]
@s-rog Something that should be more clear in the documentation? After a lot of trouble I understand that now but it was not clear from the beginning.

**Comment by gonzachiar** [comment]
@s-rog I don't understand why you are doing the `self.log_metrics(metrics, 0)` in line 175, why is this needed? Because it messes up some plots if you call `log_hyperparams` after the training.
